x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

```diff
@@ -490,7 +492,87 @@ static XContentBuilder buildUpdatedDocument(
 
         addCreator(builder, authentication);
 
-        return builder.endObject();
+        return new ApiKeyDocBuilderWithNoopFlag(
+            builder.endObject(),
+            isNoop(currentApiKeyDoc, targetDocVersion, authentication, request, userRoles)
+        );
+    }
+
+    private boolean isNoop(
+        final ApiKeyDoc apiKeyDoc,
+        final Version targetDocVersion,
+        final Authentication authentication,
+        final UpdateApiKeyRequest request,
+        final Set<RoleDescriptor> userRoles
+    ) {
+        if (apiKeyDoc.version != targetDocVersion.id) {
+            return false;
+        }
+
+        final Map<String, Object> currentCreator = apiKeyDoc.creator;
+        final var user = authentication.getEffectiveSubject().getUser();
+        final var sourceRealm = authentication.getEffectiveSubject().getRealm();
+        if (false == (Objects.equals(user.principal(), currentCreator.get("principal"))
+            && Objects.equals(user.fullName(), currentCreator.get("full_name"))
+            && Objects.equals(user.email(), currentCreator.get("email"))
+            && Objects.equals(user.metadata(), currentCreator.get("metadata"))
+            && Objects.equals(sourceRealm.getName(), currentCreator.get("realm"))
+            && Objects.equals(sourceRealm.getType(), currentCreator.get("realm_type")))) {
+            return false;
+        }
+        if (sourceRealm.getDomain() != null) {
+            if (currentCreator.get("realm_domain") == null) {
+                return false;
+            }
+            @SuppressWarnings("unchecked")
+            final var currentRealmDomain = RealmDomain.fromXContent(
+                XContentHelper.mapToXContentParser(
+                    XContentParserConfiguration.EMPTY,
+                    (Map<String, Object>) currentCreator.get("realm_domain")
+                )
+            );
+            if (sourceRealm.getDomain().equals(currentRealmDomain) == false) {
+                return false;
+            }
+        } else {
+            if (currentCreator.get("realm_domain") != null) {
+                return false;
+            }
+        }
```

Comment:
In a future PR, it is worthwhile to extract some of this logic to a separate method (something like `ApiKeyDoc.getOwnerSubject`) which can be reused in other places like `validateApiKeyExpiration`.